Docs: Document TSDB block upload feature (#2281)
grafana/mimir: docs/sources/migration-guide/migrating-from-thanos-or-prometheus.md

For configuration of remote write to Grafana Mimir, refer to [Configuring Prometheus remote write]({{< relref "../operators-guide/secure/authentication-and-authorization.md#configuring-prometheus-remote-write" >}}).

## Uploading historic TSDB blocks to Grafana Mimir

Grafana Mimir supports uploading of historic TSDB blocks, notably from Prometheus.
In order to enable this functionality, either for all tenants or a specific one, refer to
[Configuring TSDB block upload]({{< relref "../operators-guide/configure/configuring-tsdb-block-upload.md" >}}).

Prometheus stores TSDB blocks in the path specified in the `--storage.tsdb.path` flag.

To find all block directories in the TSDB `<STORAGE TSDB PATH>`, run the following command:

```bash
find <STORAGE TSDB PATH> -name chunks -exec dirname {} \;
```

Grafana Mimir supports multiple tenants and stores blocks per tenant. With multi-tenancy disabled, there
is a single tenant called `anonymous`.

Use Grafana mimirtool to upload each block, e.g. identified by the previous command, to Grafana Mimir:

```bash

> mimirtool backfill --user=<tenant> --address=http://<mimir-hostname> --id=<tenant> <block1> <block2>...

Reviewer: Already addressed :)

```

**Note**: If you need to authenticate against Grafana Mimir, you can provide an API key via the `--key` flag
(e.g., `--key=$(cat token.txt)`).

> Grafana Mimir will perform some sanitization and validation of each block's metadata.

Reviewer: Given that we cannot use `mimirtool backfill` to upload Thanos blocks at all, we should mention that limitation here as well, and also mention a workaround: direct upload of blocks to Mimir blocks bucket. Reason is that this document is about "Migrating from Thanos or Prometheus..." but if we only mention `mimirtool backfill`, it makes migrating from Thanos impossible. Yet we still talk about Thanos later in this doc.

Author: Done, PTAL.

Author: Done, PTAL.

For configuration of remote write to Grafana Mimir, refer to [Configuring Prometheus remote write]({{< relref "../operators-guide/secure/authentication-and-authorization.md#configuring-prometheus-remote-write" >}}).

## Uploading historic TSDB blocks to Grafana Mimir

Grafana Mimir supports uploading of historic TSDB blocks, notably from Prometheus.
In order to enable this functionality, either for all tenants or a specific one, refer to
[Configuring TSDB block upload]({{< relref "../operators-guide/configure/configuring-tsdb-block-upload.md" >}}).

Prometheus stores TSDB blocks in the path specified in the `--storage.tsdb.path` flag.

To find all block directories in the TSDB `<STORAGE TSDB PATH>`, run the following command:

```bash
find <STORAGE TSDB PATH> -name chunks -exec dirname {} \;
```

Grafana Mimir supports multiple tenants and stores blocks per tenant. With multi-tenancy disabled, there
is a single tenant called `anonymous`.

Use Grafana mimirtool to upload each block, e.g. identified by the previous command, to Grafana Mimir:

```bash
mimirtool backfill --address=http://<mimir-hostname> --id=<tenant> <block1> <block2>...
```

> **Note**: If you need to authenticate against Grafana Mimir, you can provide an API key via the `--key` flag

Reviewer 2: ```suggestion
> **Note**: If you need to authenticate against Grafana Mimir, you can provide an API key via the `--key` flag
```

> (e.g., `--key=$(cat token.txt)`).

Reviewer 2: ```suggestion
for example `--key=$(cat token.txt)`.
```

> Grafana Mimir will perform some sanitization and validation of each block's metadata.

Reviewer 2: ```suggestion
Grafana Mimir performs some sanitization and validation of each block's metadata.
```

As part of this, it will reject Thanos blocks due to unsupported labels.

> The workaround, if you need to upload Thanos blocks, is to upload the blocks directly to the

Reviewer 2: ```suggestion
As a workaround, if you need to upload Thanos blocks, is to upload the blocks directly to the
```

Author: @osg-grafana the edit should probably be:

> As a workaround, if you need to upload Thanos blocks, upload the blocks directly to the

> Grafana Mimir blocks bucket, prefixed by `<tenant>/<block ID>/`.

Reviewer 3: I disagree on this. There's a reason why we reject Thanos blocks with external labels. We shouldn't recommend to just copy-paste them in the storage. Instead, we should link to "Considerations on Thanos specific features" below.

Author: Thank you @pracucci :)

Reviewer: We have entire section dedicated to Thanos blocks, and workarounds to get them working. If we completely discourage people from uploading Thanos blocks to Mimir storage, we can just say it's not supported, and remove all that text.

Author: > We have entire section dedicated to Thanos blocks, and workarounds to get them working. If we completely discourage people from uploading Thanos blocks to Mimir storage, we can just say it's not supported, and remove all that text.

@pracucci what's the consensus?

For configuration of remote write to Grafana Mimir, refer to [Configuring Prometheus remote write]({{< relref "../operators-guide/secure/authentication-and-authorization.md#configuring-prometheus-remote-write" >}}).

## Uploading historic TSDB blocks to Grafana Mimir

> Grafana Mimir supports uploading of historic TSDB blocks, also from other systems such as Prometheus and Thanos.

Reviewer: Let's only mention Prometheus, since we cannot upload Thanos blocks via `mimirtool backfill`.

Author: Done.

[Configuring TSDB block upload]({{< relref "../operators-guide/configure/configuring-tsdb-block-upload.md" >}}).

Prometheus stores TSDB blocks in the path specified in the `--storage.tsdb.path` flag.

To find all block directories in the TSDB `<STORAGE TSDB PATH>`, run the following command:

```bash
find <STORAGE TSDB PATH> -name chunks -exec dirname {} \;
```

Grafana Mimir supports multiple tenants and stores blocks per tenant. With multi-tenancy disabled, there
is a single tenant called `anonymous`.

Use Grafana mimirtool to upload each block, e.g. identified by the previous command, to Grafana Mimir:

```bash
mimirtool backfill --user=<tenant> --address=http://<mimir-hostname> --id=<tenant> <block1> <block2>...
```

**Note**: If you need to authenticate against Grafana Mimir, you can provide an API key via the `--key` flag
(e.g., `--key=$(cat token.txt)`).

Grafana Mimir will perform some conversions of each block, if they are from another system
(e.g. Prometheus or Thanos). Only one label in meta.json is kept: `__compactor_shard_id__`.
The following deprecated labels are stripped:

- `__org_id__`
- `__ingester_id__`
- `__shard_id__`

> 

Reviewer: This may be too much information.
- Only Thanos, Cortex or Mimir blocks can have labels. Prometheus blocks have no labels.
- Labels listed as "Deprecated" labels are really deprecated only by Mimir. Cortex still uses `__org_id__` or `__ingester_id__`.

I'd suggest to drop this entire section, and promote block upload only for Prometheus blocks. *ALL* Thanos block will be rejected by this API, since Thanos blocks *must* use external labels. (We can mention that below in "Considerations on Thanos specific features").

For configuration of remote write to Grafana Mimir, refer to [Configuring Prometheus remote write]({{< relref "../operators-guide/securing/authentication-and-authorization.md#configuring-prometheus-remote-write" >}}).

## Uploading historic TSDB blocks to Grafana Mimir

Grafana Mimir supports uploading of historic TSDB blocks, also from other systems such as Prometheus and Thanos.
In order to enable this functionality, either for all tenants or a specific one, refer to
[Configuring TSDB block upload]({{< relref "../operators-guide/configuring/configuring-tsdb-block-upload.md" >}}).

Prometheus stores TSDB blocks in the path specified in the `--storage.tsdb.path` flag.

To find all block directories in the TSDB `<STORAGE TSDB PATH>`, run the following command:

```bash
find <STORAGE TSDB PATH> -name chunks -exec dirname {} \;
```

Grafana Mimir supports multiple tenants and stores blocks per tenant. With multi-tenancy disabled, there
is a single tenant called `anonymous`.

Use Grafana mimirtool to upload each block, e.g. identified by the previous command, to Grafana Mimir:

```bash

> mimirtool backfill --user=<tenant> --key=$(cat token.txt) --address=http://<mimir-hostname> --id=<tenant> <block1> <block2>...

Author: Thanks, done.

Grafana Mimir supports multiple tenants and stores blocks per tenant. With multi-tenancy disabled, there
is a single tenant called `anonymous`.

- For AWS S3, use the `aws` tool in the following command:
Use Grafana mimirtool to upload each block, e.g. identified by the previous command, to Grafana Mimir:

```bash
aws s3 cp <DIRECTORY> s3://<TENANT>/<DIRECTORY>
mimirtool backfill --user=<tenant> --key=$(cat token.txt) --address=http://<mimir-hostname> --id=<tenant> <block1> <block2>...
```

- For Google Cloud Storage (GCS), use the `gsutil` tool in the following command:

```bash
gsutil -m cp -r <DIRECTORY> gs://<TENANT>/<DIRECTORY>
```

> ## Block metadata

Author: @pracucci sorry, I just resolved a merge conflict heavy-handedly since I'm trying to concentrate on mimirtool work at the moment. I'm going to fix it up before moving the PR out of the WIP stage :)

> ## Block metadata

Author: I added it back.

Each block has a `meta.json` metadata file that is used by Grafana Mimir, Prometheus, and Thanos to identify the block contents.
Each project has its own metadata conventions.

In the Grafana Mimir 2.1 (or earlier) release, the ingesters added an external label to the `meta.json` file to identify the tenant that owns the block.

In the Grafana Mimir 2.2 (or later) release, blocks no longer have a label that identifies the tenant.

> **Note**: Blocks from Prometheus do not have any external labels stored in them; only blocks from Thanos use labels.

> ## Considerations on Thanos specific features

Author: @pracucci sorry, I just resolved a merge conflict heavy-handedly. I'm going to fix it up before moving the PR out of the WIP stage :)

> Grafana Mimir will perform the necessary conversions of each block, if they are from another system

Author: @pracucci good point, I'll verify what kind of conversion we're doing.

> Grafana Mimir will perform the necessary conversions of each block, if they are from another system

Author: @pracucci I rewrote this part a bit, in order to precisely mirror what is done of meta.json conversion and sanitization. I don't say that we validate blocks, since it's not yet implemented. We should document also that (block validation), once it's implemented.
